draft-ietf-gnap-resource-servers.md

```diff
@@ -0,0 +1,502 @@
+
+While it is possible to use a [token format]{#token-format} that allows for the
+RS to generate its own secondary token,
+the AS can allow the RS to request this secondary access token using the same
+process used by the original client instance to request the primary access token. Since the
+RS is acting as its own client instance from the perspective of GNAP, this process
+uses the same grant endpoint, request structure, and response structure as a client
+instance's request.
+
+~~~
++--------+       +-------+       +------+       +-------+
+| Client |--(1)->|  RS1  |       |  AS  |       |  RS2  |
+|Instance|       |       |--(2)->|      |       |       |
+|        |       |       |<-(3)--|      |       |       |
+|        |       |       |       +------+       |       |
+|        |       |       |                      |       |
+|        |       |       |-----------(4)------->|       |
+|        |       |       |<----------(5)--------|       |
+|        |<-(6)--|       |                      |       |
++--------+       +-------+                      +-------+
+~~~
+
+1. The client instance calls RS1 with an access token.
+
+2. RS1 presents that token to the AS to get a derived token
+    for use at RS2. RS1 indicates that it has no ability
+    to interact with the RO. Note that
+    RS1 signs its request with its own key, not the token's
+    key or the client instance's key.
+
+3. The AS returns a derived token to RS1 for use at RS2.
+
+4. RS1 calls RS2 with the token from (3).
+
+5. RS2 fulfills the call from RS1.
+
+6. RS1 fulfills the call from the original client instance.
+
+
+If the RS needs to derive a token from one presented to it, it can
+request one from the AS by making a token request as described in
+{{I-D.ietf-gnap-core-protocol}} and presenting the existing access token's
+value in the "existing_access_token" field.
+
+Since the RS is acting as a client instance, 
+the RS MUST identify itself with its own key in the `client` field and sign the
+request just as any client instance would.
+
+\[\[ [See issue #116](https://github.com/ietf-wg-gnap/gnap-core-protocol/issues/116) \]\]
+
+~~~
+POST /tx HTTP/1.1
+Host: server.example.com
+Content-Type: application/json
+Detached-JWS: ejy0...
+
+{
+    "access_token": {
+        "access": [
+            {
+                "actions": [
+                    "read",
+                    "write",
+                    "dolphin"
+                ],
+                "locations": [
+                    "https://server.example.net/",
+                    "https://resource.local/other"
+                ],
+                "datatypes": [
+                    "metadata",
+                    "images"
+                ]
+            },
+            "dolphin-metadata"
+        ]
+    },
+    "client": "7C7C4AZ9KHRS6X63AJAO",
+    "existing_access_token": "OS9M2PMHKUR64TB8N6BW7OZB8CDFONP219RP1LT0"
+}
+~~~
+
+The AS responds with a token for the downstream RS2 as described in 
```

Comment:
I would suggest omitting the `downstream` word for https://github.com/ietf-wg-gnap/gnap-core-protocol/pull/246#issuecomment-819066135 to make this section more generic. An RS 1 may call RS2 that belongs to a completely different sec domain. RS 2 may also trust only to its own AS 2... and there is some form of trust established between AS 1 and AS 2. The same applies to the name of this section (Deriving a downstream token). I do hope it makes sense.

Reply:
Fair point, I think the current section is tightly focused on the RS going back to the same issuing AS that the original token came from -- this doesn't have to be the case, of course, but for the cases where it's not the issuing AS then as you point out there needs to be some kind of trust happening between the two AS's such that the second AS can make any sense of things. I am a little afraid of the model becoming overly abstract and not giving developers enough to hook on to, but I think it's worth pursuing. I'm going to suggest that if the WG does accept this PR then we continue that conversation in a new issue dedicated to that topic.